modules/stochastic_tools/src/utils/Calculators.C

```diff
@@ -168,27 +146,74 @@ Ratio::compute(const std::vector<Real> & data, bool is_distributed) const
 
   if (is_distributed)
   {
-    _communicator.min(local_min);
-    _communicator.max(local_max);
+    this->_communicator.min(local_min);
+    this->_communicator.max(local_max);
   }
 
   return local_min != 0. ? local_max / local_min : 0.;
 }
 
 // L2NORM //////////////////////////////////////////////////////////////////////////////////////////
-L2Norm::L2Norm(const libMesh::ParallelObject & other) : Calculator(other) {}
-
-Real
-L2Norm::compute(const std::vector<Real> & data, bool is_distributed) const
+template <typename InType, typename OutType>
+OutType
+L2Norm<InType, OutType>::compute(const InType & data, bool is_distributed) const
 {
   Real local_sum =
       std::accumulate(data.begin(), data.end(), 0., [](Real running_value, Real current_value) {
         return running_value + std::pow(current_value, 2);
       });
 
   if (is_distributed)
-    _communicator.sum(local_sum);
+    this->_communicator.sum(local_sum);
 
   return std::sqrt(local_sum);
 }
-} // Statistics namespace
+
+template class Calculator<std::vector<Real>, Real>;
+template class Mean<std::vector<Real>, Real>;
+template class Max<std::vector<Real>, Real>;
+template class Min<std::vector<Real>, Real>;
+template class Sum<std::vector<Real>, Real>;
+template class StdDev<std::vector<Real>, Real>;
+template class StdErr<std::vector<Real>, Real>;
+template class Ratio<std::vector<Real>, Real>;
+template class L2Norm<std::vector<Real>, Real>;
+
+/*
+ * Free function for building a const Calculator object for use by Statistics object
+ */
+std::unique_ptr<const Calculator<std::vector<Real>, Real>>
+makeCalculator(const MooseEnumItem & item, const libMesh::ParallelObject & other)
+{
+  if (item == "min")
+    return libmesh_make_unique<const Min<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "max")
+    return libmesh_make_unique<const Max<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "sum")
+    return libmesh_make_unique<const Sum<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "mean" || item == "average") // average is deprecated
+    return libmesh_make_unique<const Mean<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "stddev")
+    return libmesh_make_unique<const StdDev<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "stderr")
+    return libmesh_make_unique<const StdErr<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "norm2")
+    return libmesh_make_unique<const L2Norm<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "ratio")
+    return libmesh_make_unique<const Ratio<std::vector<Real>, Real>>(other, item);
+
+  ::mooseError("Failed to create Statistics::Calculator object for ", item);
+  return nullptr;
+}
```

Comment:
This non-template version works fine. If you switch to the template version there are two or three calls that template arguments need to be added. I also attempted to set default template arguments, but that didn't help.